qi-lib/flow/core/compiler.rkt

```diff
@@ -18,8 +18,96 @@
   (define (compile-flow stx)
     #`(qi0->racket #,(optimize-flow stx)))
 
+  (define (optimization-pass stx)
+    ;; TODO: the "active" components of the expansions should be
+    ;; optimized, i.e. they should be wrapped with a recursive
+    ;; call to the optimizer
+    (syntax-parse stx
+      ;; restorative optimization for "all"
+      [((~datum thread) ((~datum amp) onex) (~datum AND))
+       #`(esc (give (curry andmap #,(compile-flow #'onex))))]
+      ;; "deforestation" for values
+      ;; (~> (pass f) (>< g)) → (>< (if f g ⏚))
+      [((~datum thread) _0 ... ((~datum pass) f) ((~datum amp) g) _1 ...)
+       #'(thread _0 ... (amp (if f g ground)) _1 ...)]
+      ;; (~> (>< g) (pass f)) → (>< (~> g (if f _ ⏚)))
```

Comment:
An observable difference in behaviors: partly the difference is arity related. In `orig`, `f` consumes each value of `(>< g)` one at a time; in `opt`, it consumes _all_ values of `(>< g)`.
```
> ,r qi
> (define-flow g (-< add1 sub1))
> (define-flow f positive?)
> (define (f* x y) (= (sub1 x) (add1 y)))
> (define (orig g f) (flow (~> (>< g) (pass f))))
> (define (opt g f) (flow (>< (~> g (if f _ ground)))))
> (apply (orig g f) (range -3 4))
1
2
3
1
4
2
> (apply (opt g f) (range -3 4))
positive?: arity mismatch;
 the expected number of arguments does not match the given number
  expected: 1
  given: 2
 [,bt for context]
> (apply (orig g f*) (range -3 4))
filter: contract violation
  expected: (any/c . -> . any/c)
  given: #<procedure:f*>
 [,bt for context]
> (apply (opt g f*) (range -3 4))
-2
-4
-1
-3
0
-2
1
-1
2
0
3
1
4
2
```

Reply:
Thanks for catching that!

Reply:
Brilliant, good idea.